### ISSUE TITLE
System metrics are offered as gauge instead of counter

### ISSUE THREAD

**Issue by rgueldenpfennig** [prometheus-net/prometheus-net#387]
I just updated to the latest 7.0.0 version of prometheus-net and noticed that many .NET runtime system telemetry counters are now registered by default.

But I'm a bit surprised that for example the `system_net_sockets_bytes_received` and `system_net_sockets_bytes_sent` metrics are registered as a gauge type. I would expect a counter here.

As far as I can see both telemetry counters use a `long` as the backing field and are [incremented over time](https://github.com/dotnet/runtime/commit/d3199700654c46994157eb4974f98a4bcb6edf76). A gauge would also decrease like the total count of threads.

Am I missing here something?

My issue is that I'm not able to utilize functions like `rate` or `increase` with a gauge metric.

**Comment by rgueldenpfennig** [comment]
So I just tried to dig into the source code and think that the reason for that behavior is handled here: https://github.com/prometheus-net/prometheus-net/blob/master/Prometheus/EventCounterAdapter.cs#L112-L138

Maybe it's not yet sufficient enough to distinguish between gauge and counter when handling an event. Will try to analyze further but would also appreciate any feedback here.

**Comment by sandersaares** [comment]
Yes, from what I recall there was some difficulty mapping from the .NET event counter universe to the Prometheus universe, so the most universal hammer for this nail seemed to be to make them all a gauge. Possibly there is a better solution but it did not seem worth the time to invest into untangling the situation.

> My issue is that I'm not able to utilize functions like rate or increase with a gauge metric.

What stops you, exactly? As far as I know, the Prometheus database will allow you to use rate() or increase() over any timeseries, no matter whether the metric that created the timeseries is declared as a counter or gauge.

**Comment by avendel** [comment]
I saw that this also applies to `System.Diagnostics.Metrics` counters, which are now registered as Gauges. I found the commit which introduces this behavior. The comments added with https://github.com/prometheus-net/prometheus-net/commit/f1e36234d436fddfd6dcbc154a13f367005be642#diff mentions:

>  We do not represent any of the "counter" style .NET meter types as counters because they may be re-created on the .NET Meters side at any time, decrementing the value!

Is this referring to a scenario where you create multiple instance for the same counter name on the .NET side? I don't think this is a problem, though, since the `System.Diagnostics.Metrics` instruments only forward observations to listeners. There is no current value stored on the .NET side. 

See e.g. https://github.com/dotnet/runtime/blob/release/7.0/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Counter.cs#L30 which eventually calls listeners at https://github.com/dotnet/runtime/blob/release/7.0/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Instrument.common.cs#L45.

If this was a problem, you would have an issue with histograms as well, right? They are basically represented as a collection of counters, one per bucket.

Or did I misunderstand the comments from https://github.com/prometheus-net/prometheus-net/commit/f1e36234d436fddfd6dcbc154a13f367005be642?

**Comment by slang25** [comment]
I have also just run into this, could we at least make it an option so that library consumers can decide what behavior they want.

I'll create a PR if there are no objections. @sandersaares are you still actively reviewing PRs? I see #410 hasn't been looked at.

**Comment by jfreeland** [comment]
just noticed this as well.  glad it's a known issue. :)

**Comment by bekir-ozturk** [comment]
Ran into this today with my custom metrics. 
Considering how different gauges and counters are, I'm surprised that this change was merged in the first place.

For anyone looking for an alternative:
- I tried [OpenTelemetry.Exporter.Prometheus.HttpListener](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Exporter.Prometheus.HttpListener). It seems to correctly capture counters, but it doesn't seem to be emitting System.Runtime metrics. There is a plan to get this done natively in the runtime itself (https://github.com/dotnet/runtime/issues/85372) which will mean we won't need to rely on exporters for this, but it isn't available today.
- Second option I can think of is to fork this repo and revert the change. I'll give this one a try.
